### ISSUE TITLE
RTCPeerConnection.close procedure: Data Channel & SCTP questions

### ISSUE THREAD

**Issue by lgrahl** [w3c/webrtc-pc#1821]
I have a bunch of questions for the closing procedure of `RTCPeerConnection.close`:

1. The `ReadyState` of the `RTCDataChannel`s is updated to `closed`. I'm assuming here this means the data channels will be closed in an abrupt manner. If you don't think this is the case, ignore the sub-steps below.
    1. Should we add a note that this means data channels will be closed abruptly and the *closing* procedure is **not** invoked?
        (I think we should)
    2. Does that mean the `onclose` event will fire? Related: https://github.com/w3c/webrtc-pc/issues/1799
        (I don't think it needs to since we have `RTCSctpTransport.onstatechange`)
    3. Is the *closed* procedure invoked (see below)? Should it fire the `RTCError` event?
     
         > 2 . Set channel's [[ReadyState]] slot to closed.
         > 3 . Fire an RTCError event at channel with errorDetail set to "data-channel-failure".
         > 4 . Fire a simple event named close at channel.

        (I believe it should invoke the *closed* procedure and reference that. And I think it should also fire since that's a non-graceful close.)
2. The `SCTPTransportState` slot is updated to `closed`. Does that mean the SCTP association [will be aborted or terminated gracefully](https://tools.ietf.org/html/draft-ietf-rtcweb-data-channel-13#section-6.2)?

(I think it should state to send an SCTP ABORT chunk and be done with it since all data channels have already been closed non-gracefully and therefore all data in-flight is lost.)

**Comment by lgrahl** [comment]
I want to add that applications can quite easily achieve closing gracefully by closing all data channels, wait for them to become *closed* and then close the peer connection.

**Comment by lgrahl** [comment]
And to add another thing: I would propose to have both closing graceful (could return a `Promise`) and non-graceful for WebRTC NV.

**Comment by taylor-b** [comment]
I agree that `RTCPeerConnection.close` should perform an abrupt closure, sending an SCTP ABORT chunk. Since `close` is synchronous (doesn't return a promise), and results in everything being closed, it can't do anything graceful without blocking the main thread or lying about things being closed.

I'm not sure about the events, though. The general policy with events seems to be "don't fire anything while closing the `RTCPeerConnection`"; there's always a step that says "If `connection`'s [[IsClosed]] slot is `true`, abort these steps."

**Comment by lgrahl** [comment]
Oh yeah, it seems I forgot this policy after 1.i. Fair point. Then the only suggestion I have for the data channel procedure is to explicitly mention that neither the *closing* procedure, nor the *closed* procedure will be invoked.

**Comment by adam-be** [comment]
I would argue that the close method doesn't abruptly close the channel as currently specified. The setting of the readyState to "closed" (referred to above) is in the list of operations that must be run when the closing procedure is *done* (to notify the script about it). So the close method only updates the readyState to "closing".

**Comment by lgrahl** [comment]
@adam-be Are you `RTCPeerConnection.close` or `RTCDataChannel.close`?

**Comment by aboba** [comment]
i. Agree that pc.close() implies an abrupt close.
ii. Don't think that `onclose` event should fire.
iii. Don't think that the `RTCError` event should fire.

**Comment by lgrahl** [comment]
Good, so there seems to be general agreement to 1.i and 2. - and 1.ii and 1.iii are moot due to the rule to not raise events due to user action.
